vhost-device-input/src/vhu_input.rs

```diff
@@ -0,0 +1,625 @@
+// VIRTIO Input Emulation via vhost-user
+//
+// Copyright 2023 Linaro Ltd. All Rights Reserved.
+// Leo Yan <leo.yan@linaro.org>
+//
+// SPDX-License-Identifier: Apache-2.0 or BSD-3-Clause
+
+use log::error;
+use nix::ioctl_read_buf;
+use std::collections::VecDeque;
+use std::{convert, io, result};
+use thiserror::Error as ThisError;
+
+use vhost::vhost_user::message::{VhostUserProtocolFeatures, VhostUserVirtioFeatures};
+use vhost_user_backend::{VhostUserBackendMut, VringRwLock, VringT};
+use virtio_bindings::bindings::virtio_config::VIRTIO_F_VERSION_1;
+use virtio_bindings::bindings::virtio_ring::{
+    VIRTIO_RING_F_EVENT_IDX, VIRTIO_RING_F_INDIRECT_DESC,
+};
+use virtio_queue::QueueT;
+use vm_memory::{ByteValued, Bytes, GuestAddressSpace, GuestMemoryAtomic, GuestMemoryMmap};
+use vmm_sys_util::epoll::EventSet;
+use vmm_sys_util::eventfd::{EventFd, EFD_NONBLOCK};
+
+use crate::input::*;
+
+type Result<T> = std::result::Result<T, VuInputError>;
+
+pub const EVENT_ID_IN_VRING_EPOLL: u64 = 3;
+
+const QUEUE_SIZE: usize = 1024;
+const NUM_QUEUES: usize = 2;
+
+const VIRTIO_INPUT_CFG_ID_NAME: u8 = 0x01;
+const VIRTIO_INPUT_CFG_ID_DEVIDS: u8 = 0x03;
+const VIRTIO_INPUT_CFG_EV_BITS: u8 = 0x11;
+const VIRTIO_INPUT_CFG_SIZE: usize = 128;
+
+const EV_SYN: u8 = 0x00;
+const EV_KEY: u8 = 0x01;
+const EV_REL: u8 = 0x02;
+const EV_ABS: u8 = 0x03;
+const EV_MSC: u8 = 0x04;
+const EV_SW: u8 = 0x05;
+
+const SYN_REPORT: u8 = 0x00;
+
+pub(crate) trait ToBytes {
+    fn to_bytes(&self) -> Vec<u8>;
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+#[repr(C)]
+pub(crate) struct VuInputConfig {
+    select: u8,
+    subsel: u8,
+    size: u8,
+    reserved: [u8; 5],
+    val: [u8; VIRTIO_INPUT_CFG_SIZE],
+}
+
+impl ToBytes for VuInputConfig {
+    fn to_bytes(&self) -> Vec<u8> {
+        let ret = [
+            self.select.as_slice(),
+            self.subsel.as_slice(),
+            self.size.as_slice(),
+            self.reserved.as_slice(),
+            self.val.as_slice(),
+        ]
+        .concat();
+        ret
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+#[repr(C)]
+pub(crate) struct VuInputEvent {
+    ev_type: u16,
+    code: u16,
+    value: u32,
```

Comment:
Where is this struct defined in a spec? Should it be always Little Endian or always Big Endian?